File: ED del Pedro/ExamenesED/Febrero2020/code/haskell/src/DataStructures/Graph/FordFulkerson.hs
-------------------------------------------------------------------------------
-- Ford-Fulkerson Algorithm. Maximal flow for a weighted directed graph.
--
-- Student's name:
-- Student's group:
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.FordFulkerson where

import Data.List  ((\\))
import DataStructures.Graph.WeightedDiGraph
import DataStructures.Graph.WeightedDiGraphBFT

maxFlowPath :: Path (WDiEdge a Integer) -> Integer
maxFlowPath hs = minimum (map getWeight hs)

getWeight :: WDiEdge a Integer -> Integer
getWeight (E x w y) = w

updateEdge ::(Eq a) => a -> a -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
updateEdge x y w [] = [E x w y]
updateEdge x y w (h : hs)
    | (x, y) == getOrigDest h = (updateWeight w h) ++ hs
    | otherwise = h : (updateEdge x y w hs)

getOrigDest :: WDiEdge a Integer -> (a, a)
getOrigDest (E x w y) = (x, y)

updateWeight :: Integer -> WDiEdge a Integer -> [WDiEdge a Integer]
updateWeight w (E x w' y)
    | w + w' == 0 = []
    | otherwise = [E x (w + w') y]

updateEdges :: (Eq a) => Path (WDiEdge a Integer) -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
updateEdges [] _ edges = edges
updateEdges (p : ps) w edges = updateEdges ps w (updateEdge x y w edges)
    where
        (x, y) = getOrigDest p

addFlow :: (Eq a) => a -> a -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
addFlow x y p [] = [E x p y]
addFlow x y p (s : ss)
    | (x, y) == getOrigDest s = (E x (p + w) y) : ss
    | (y, x) == getOrigDest s && p == w = ss
    | (y, x) == getOrigDest s = if w < p then (E x (p - w) y) : ss else (E y (w - p) x) : ss
    | otherwise = s : (addFlow x y p ss)
        where w = getWeight s

addFlows :: (Eq a) => Path (WDiEdge a Integer) -> Integer -> [WDiEdge a Integer] -> [WDiEdge a Integer]
addFlows [] _ sol = sol
addFlows (p : ps) w sol = addFlows ps w (addFlow x y w sol)
    where
        (x, y) = getOrigDest p

fordFulkerson :: (Ord a) => (WeightedDiGraph a Integer) -> a -> a -> [WDiEdge a Integer]
fordFulkerson g src dst = aux g src dst [] (bftPathTo g src dst)
    where
        aux _ _ _ sol Nothing = sol
        aux wdg src dst sol (Just path) =  aux newWdg src dst (addFlows path mf sol) (bftPathTo newWdg src dst)
            where
                mf = maxFlowPath path
                reverseEdges = reversePath path
                edges = (updateEdges reverseEdges mf (updateEdges path (-mf) (weightedDiEdges wdg)))
                newWdg = mkWeightedDiGraphEdges (vertices wdg) edges

reversePath :: Path (WDiEdge a Integer) -> Path (WDiEdge a Integer)
reversePath [] = []
reversePath (e : es) = reversePath es ++ [reverseEdge e]

reverseEdge :: WDiEdge a Integer -> WDiEdge a Integer
reverseEdge (E x w y) = E y w x

--reversePath [E 'a' 5 'b', E 'b' 2 'c', E 'c' 3 'd']

maxFlow :: (Ord a) => [WDiEdge a Integer] -> a -> Integer
maxFlow [] src = 0
maxFlow (s : ss) src
    | x == src = w + maxFlow ss src
    | otherwise = maxFlow ss src
        where 
            (x, y) = getOrigDest s
            w = getWeight s

maxFlowMinCut :: (Ord a) => (WeightedDiGraph a Integer) -> a -> a -> [a] -> Integer
maxFlowMinCut g src dst set
    | elem src set = sumaKK' - sumaK'K
    | otherwise = sumaK'K - sumaKK'
        where
            edges = fordFulkerson g src dst
            sumaKK' = sum (map (\ (E x w y) -> if elem x set && elem y dif then w else 0) edges)
            sumaK'K = sum (map (\ (E x w y) -> if elem y set && elem x dif then w else 0) edges)
            dif = (vertices g) \\ set



-- A partir de aquí hasta el final
-- SOLO para alumnos a tiempo parcial 
-- sin evaluación continua

localEquilibrium :: (Ord a) => WeightedDiGraph a Integer -> a -> a -> Bool
localEquilibrium = undefined

sourcesAndSinks :: (Eq a) => WeightedDiGraph a b -> ([a],[a])
sourcesAndSinks = undefined

unifySourceAndSink :: (Eq a) => WeightedDiGraph a Integer -> a -> a -> WeightedDiGraph a Integer
unifySourceAndSink = undefined

File: ED del Pedro/ExamenesED/Febrero2020/code/java/src/dataStructures/graph/FordFulkerson.java
/**
 * Student's name:
 *
 * Student's group:
 */

package dataStructures.graph;

import dataStructures.list.LinkedList;
import dataStructures.list.List;
import dataStructures.set.Set;
import dataStructures.tuple.Tuple2;

import java.lang.reflect.WildcardType;

public class FordFulkerson<V> {
    private WeightedDiGraph<V,Integer> g; // Initial graph
    private List<WDiEdge<V,Integer>> sol; // List of edges representing maximal flow graph
    private V src; 			  // Source
    private V dst; 		  	  // Sink

    /**
     * Constructors and methods
     */

    public static <V> int maxFlowPath(List<WDiEdge<V,Integer>> path) {
        int min = path.get(0).getWeight();

        for (WDiEdge<V,Integer> wde : path){
            int w = wde.getWeight();
            if(w < min){
                min = w;
            }
        }

        return min;
    }

    public static <V> List<WDiEdge<V,Integer>> updateEdge(V x, V y, Integer p, List<WDiEdge<V,Integer>> edges) {
        boolean found = false;
        int i = 0;
        while (! found && i < edges.size()){
            WDiEdge<V,Integer> edge = edges.get(i);
            if(edge.getSrc().equals(x) && edge.getDst().equals(y)){
                if(edge.getWeight() + p == 0){
                    edges.remove(i);
                }else {
                    edges.set(i, new WDiEdge<>(x, p + edge.getWeight(), y));
                }
                found = true;
            }
            i ++;
        }

        if(! found){
            edges.append(new WDiEdge<>(x, p, y));
        }

        return edges;
    }

    public static <V> List<WDiEdge<V,Integer>> updateEdges(List<WDiEdge<V,Integer>> path, Integer p, List<WDiEdge<V,Integer>> edges) {

        if(path != null){
            for (WDiEdge<V, Integer> wde : path){
                edges = updateEdge(wde.getSrc(), wde.getDst(), p, edges);
            }
        }
        return edges;
    }

    public static <V> List<WDiEdge<V,Integer>> addFlow(V x, V y, Integer p, List<WDiEdge<V,Integer>> sol) {
        boolean found = false;
        int i = 0;

        while (! found && i < sol.size()){
            WDiEdge<V,Integer> edge = sol.get(i);

            if(edge.getSrc().equals(x) && edge.getDst().equals(y)){
                sol.set(i,new WDiEdge<>(x, p + edge.getWeight(), y));
                found = true;
            }else if(edge.getSrc().equals(y) && edge.getDst().equals(x)){
                if(edge.getWeight().equals(p)){
                    sol.remove(i);
                }else if(edge.getWeight() < p){
                    sol.set(i,new WDiEdge<>(x, p - edge.getWeight(), y));
                }else {
                    sol.set(i,new WDiEdge<>(x, edge.getWeight() - p, y));
                }
                found = true;
            }
            i ++;
        }

        if (! found){
            sol.append(new WDiEdge<>(x, p, y));
        }

        return sol;
    }

    public static <V> List<WDiEdge<V,Integer>> addFlows(List<WDiEdge<V,Integer>> path, Integer p, List<WDiEdge<V,Integer>> sol) {

        if(path != null){
            for (WDiEdge<V, Integer> wde : path){
                sol = addFlow(wde.getSrc(), wde.getDst(), p, sol);
            }
        }
        return sol;
    }

    public FordFulkerson(WeightedDiGraph<V,Integer> g, V src, V dst) {
        List<WDiEdge<V, Integer>> sol = new LinkedList<>();
        WeightedDiGraph<V, Integer> wdg = g;
        WTraversal<V, Integer> paths = new WeightedBreadthFirstTraversal<>(wdg, src);
        List<WDiEdge<V, Integer>> path = paths.pathTo(dst);

        while (path != null){
            int mf = maxFlowPath(path);
            List<WDiEdge<V, Integer>> edges = wdg.wDiEdges();
            edges = updateEdges(path, (- mf), edges);
            WTraversal<V, Integer> revPaths = new WeightedBreadthFirstTraversal<>(wdg, dst);
            List<WDiEdge<V, Integer>> revPath = revPaths.pathTo(src);
            edges = updateEdges(revPath, mf, edges);
            wdg = new WeightedDictionaryDiGraph<>(wdg.vertices(), edges);
            sol = addFlows(path, mf, sol);
            paths = new WeightedBreadthFirstTraversal<>(wdg, src);
            path = paths.pathTo(dst);
        }

        this.g = g;
        this.src = src;
        this.dst = dst;
        this.sol = sol;

    }

    public int maxFlow() {
        int max = 0;

        for (WDiEdge<V, Integer> wde : sol){
            if(wde.getSrc().equals(src)){
                max += wde.getWeight();
            }
        }
        return max;
    }

    public int maxFlowMinCut(Set<V> set) {
        int sum1 = 0;
        int sum2 = 0;
        int max = 0;
        List<WDiEdge<V, Integer>> edges = g.wDiEdges();

        for (WDiEdge<V, Integer> edge : edges){
            if(set.isElem(edge.getSrc()) && ! set.isElem(edge.getDst())){
                sum1 += edge.getWeight();
            }else if(set.isElem(edge.getDst()) && ! set.isElem(edge.getSrc())){
                sum2 += edge.getWeight();
            }
        }
        if(set.isElem(src)){
            max = sum1 - sum2;
        }else {
            max = sum2 - sum1;
        }
        return max;
    }

    /**
     * Provided auxiliary methods
     */
    public List<WDiEdge<V, Integer>> getSolution() {
        return sol;
    }

    /**********************************************************************************
     * A partir de aquí SOLO para estudiantes a tiempo parcial sin evaluación continua.
     * ONLY for part time students.
     * ********************************************************************************/

    public static <V> boolean localEquilibrium(WeightedDiGraph<V,Integer> g, V src, V dst) {
        // TO DO
        return false;
    }
    public static <V,W> Tuple2<List<V>,List<V>> sourcesAndSinks(WeightedDiGraph<V,W> g) {
        // TO DO
        return null;
    }

    public static <V> void unifySourceAndSink(WeightedDiGraph<V,Integer> g, V newSrc, V newDst) {
        // TO DO
    }
}
